sinks/attribution/attribution.go

```diff
@@ -0,0 +1,346 @@
+			s.recordMetric(metric)
+		}
+	}
+
+	if s.log.Level >= logrus.DebugLevel {
+		for k, v := range s.attributionData {
+			s.log.Debug(fmt.Sprintf("%s - %d", k, v.Sketch.Estimate()))
+		}
+	}
+
+	csv, err := encodeAttributionDataCSV(s.attributionData, s.schemaVersion)
+	if err != nil {
+		s.log.WithFields(logrus.Fields{
+			logrus.ErrorKey: err,
+			"metrics":       len(metrics),
+		}).Error("Could not marshal metrics before posting to s3")
+		return err
+	}
+
+	err = s.s3Post(csv)
+	if err != nil {
+		s.log.WithFields(logrus.Fields{
+			logrus.ErrorKey: err,
+			"metrics":       len(metrics),
+		}).Error("Error posting to s3")
+		return err
+	}
+
+	s.log.WithField("metrics", len(metrics)).Debug("Completed flush to s3")
+	spanTags := map[string]string{"sink": s.Name()}
+	span.Add(ssf.Timing(sinks.MetricKeyMetricFlushDuration, time.Since(flushStart), time.Nanosecond, spanTags))
+	return nil
+}
+
+// timeseriesDigest takes in metric metadata and derives a digest that can be
+// used to tell if multiple InterMetrics correspond to the same
+// TimeseriesGroup, for tallying purposes
+func timeseriesDigest(name string, tags []string) uint32 {
+	// Workers have some notion of timeseries uniqueness, but one that isn't
+	// quite rigorous enough to perform timeseries cardinality computation
+	// accurately:
+	// https://github.com/stripe/veneur/blob/531ffab0dc9bc105a69fa9106023ddc36b49591e/samplers/parser.go#L38-L60
+	// DogStatsD tags can be ordered arbitrarily, but (*UDPMetric) UpdateTags
+	// does not take into account that (foo:val, bar:val) and
+	// (bar:val, foo:val) are semantically identical. However, it's a more
+	// involved refactor to make Tags order agnostic, so in the meantime we
+	// have this hack so we can accurately tally timeseries:
+	sort.Strings(tags)
+
+	h := fnv1a.Init32
+	h = fnv1a.AddString32(h, name)
+	for _, tag := range tags {
+		h = fnv1a.AddString32(h, tag)
+	}
+	return h
+}
+
+// timeseriesIDs takes metric metadata and the name of a tag that can be used
+// to derive a TimeseriesGroup owner, and returns (1) an identifier that
+// ultimately corresponds to a grouping of Timeseries to be tallied and (2) an
+// owner corresponding to said group
+//
+// If an ownerKey is not specified, timeseriesIDs returns the name
+// of the metric and a blank owner
+func timeseriesIDs(name string, tags []string, ownerKey string) (groupID, ownerID string) {
+	var b strings.Builder
+	b.WriteString(name)
+
+	if ownerKey != "" {
+		// Find the tag matching ownerKey and add its value to the Group ID
+		for _, tag := range tags {
+			tagSplit := strings.SplitN(tag, ":", 2)
+			if len(tagSplit) == 2 && tagSplit[0] == ownerKey {
+				ownerID = tagSplit[1]
+				b.WriteString(ownerID)
+			}
+		}
+	}
+
+	groupID = b.String()
+	return
+}
+
+// recordMetric tallies an InterMetric, creating a TimeseriesGroup in the process
+func (s *AttributionSink) recordMetric(metric samplers.InterMetric) {
+	tags := metric.Tags
```

Comment:
this will still modify `metrics.Tags` since we're not doing a deep copy, is that fine?